nexus/db-queries/src/db/datastore/rack.rs

```diff
@@ -401,6 +404,46 @@ impl DataStore {
 
         Ok(())
     }
+
+    pub async fn nexus_external_addresses(
+        &self,
+        opctx: &OpContext,
+    ) -> Result<Vec<IpAddr>, Error> {
+        opctx.authorize(authz::Action::Read, &authz::DNS_CONFIG).await?;
+
+        use crate::db::schema::external_ip::dsl as extip_dsl;
+        use crate::db::schema::nexus_service::dsl as nexus_dsl;
+        type TxnError = TransactionError<()>;
+        self.pool_authorized(opctx)
+            .await?
+            .transaction_async(|conn| async move {
+                // This is the rare case where we want to allow table scans.
+                // There must not be enough Nexus instances for this to be a
+                // problem.  If there were, we couldn't fit them in DNS anyway.
+                let sql = crate::db::queries::ALLOW_FULL_TABLE_SCAN_SQL;
+                conn.batch_execute_async(sql).await?;
+                Ok(extip_dsl::external_ip
+                    .filter(
+                        extip_dsl::id.eq_any(
+                            nexus_dsl::nexus_service
+                                .select(nexus_dsl::external_ip_id),
+                        ),
+                    )
```

Comment:
If you mean: could we avoid tripping CockroachDB's table scan check by using the explicit INNER JOIN syntax, I don't think so?  It's too smart for that (which is good, I think).  I tried this under `omicron-dev run-all`:

```
root@127.0.0.1:58594/omicron> \x
root@127.0.0.1:58594/omicron> select * from service;
-[ RECORD 1 ]
id            | 1bb51dc9-0cfd-4306-b972-94dae3652b00
time_created  | 2023-04-24 21:28:17.338888+00
time_modified | 2023-04-24 21:28:17.338888+00
sled_id       | b6d65341-167c-41df-9b5c-41cded99c229
ip            | ::1
port          | 12221
kind          | nexus
-[ RECORD 2 ]
id            | 24b8195d-dcc5-4bf0-b5ee-399b7ff1c70a
time_created  | 2023-04-24 21:28:17.302978+00
time_modified | 2023-04-24 21:28:17.302978+00
sled_id       | b6d65341-167c-41df-9b5c-41cded99c229
ip            | ::1
port          | 40713
kind          | internal_dns_config
-[ RECORD 3 ]
id            | 8d7c00fd-45d8-4524-88d2-a65d2230074f
time_created  | 2023-04-24 21:28:17.334491+00
time_modified | 2023-04-24 21:28:17.334491+00
sled_id       | b6d65341-167c-41df-9b5c-41cded99c229
ip            | ::1
port          | 34570
kind          | external_dns_config


Time: 2ms total (execution 1ms / network 1ms)

root@127.0.0.1:58594/omicron> select * from external_ip;
-[ RECORD 1 ]
id               | 96e2738f-579f-4baf-9e17-54108668c63b
name             | NULL
description      | NULL
time_created     | 2023-04-24 21:28:17.342471+00
time_modified    | 2023-04-24 21:28:17.342471+00
time_deleted     | NULL
ip_pool_id       | 1bd76090-22f5-4019-8ab6-b6844eba3b0d
ip_pool_range_id | 9f234a00-4d60-48df-9f71-84c4d31be28e
instance_id      | NULL
kind             | service
ip               | 127.0.0.1
first_port       | 0
last_port        | 65535


Time: 2ms total (execution 1ms / network 0ms)

root@127.0.0.1:58594/omicron> select * from nexus_service;
-[ RECORD 1 ]
id             | 1bb51dc9-0cfd-4306-b972-94dae3652b00
external_ip_id | 96e2738f-579f-4baf-9e17-54108668c63b


Time: 1ms total (execution 1ms / network 0ms)

root@127.0.0.1:58594/omicron> set disallow_full_table_scans = on; set large_full_scan_rows = 0;
SET

Note: timings for multiple statements on a single line are not supported. See https://go.crdb.dev/issue-v/48180/v22.1.

root@127.0.0.1:58594/omicron> select * from external_ip INNER JOIN nexus_service ON external_ip.id = nexus_service.external_ip_id;
ERROR: query `select * from external_ip INNER JOIN nexus_service ON external_ip.id = nexus_service.external_ip_id` contains a full table/index scan which is explicitly disallowed
SQLSTATE: P0003
HINT: try overriding the `disallow_full_table_scans` or increasing the `large_full_scan_rows` cluster/session settings
root@127.0.0.1:58594/omicron> select * from nexus_service INNER JOIN external_ip ON nexus_service.external_ip_id = external_ip.id;
ERROR: query `select * from nexus_service INNER JOIN external_ip ON nexus_service.external_ip_id = external_ip.id` contains a full table/index scan which is explicitly disallowed
SQLSTATE: P0003
HINT: try overriding the `disallow_full_table_scans` or increasing the `large_full_scan_rows` cluster/session settings
root@127.0.0.1:58594/omicron> 
```

Reply:
Ah ok. I do still have a slight preference for a JOIN here. If you don't mind, does `EXPLAIN ANALYZE` show any big difference here?

Reply:
Once CockroachDB has collected stats, in my simple case, the two explicit INNER JOIN queries are the same:

```
root@127.0.0.1:60981/omicron> EXPLAIN select * from nexus_service INNER JOIN external_ip ON nexus_service.external_ip_id = external_ip.id;
                                       info
----------------------------------------------------------------------------------
  distribution: local
  vectorized: true

  • lookup join
  │ estimated row count: 1
  │ table: external_ip@external_ip_pkey
  │ equality: (external_ip_id) = (id)
  │ equality cols are key
  │
  └── • scan
        estimated row count: 1 (100% of the table; stats collected 1 minute ago)
        table: nexus_service@nexus_service_pkey
        spans: FULL SCAN
(13 rows)


Time: 3ms total (execution 2ms / network 0ms)

root@127.0.0.1:60981/omicron> EXPLAIN select * from external_ip INNER JOIN nexus_service ON external_ip.id = ne                        id;
                                       info
-----------------------------------------------------------------------------------
  distribution: local
  vectorized: true

  • lookup join
  │ estimated row count: 1
  │ table: external_ip@external_ip_pkey
  │ equality: (external_ip_id) = (id)
  │ equality cols are key
  │
  └── • scan
        estimated row count: 1 (100% of the table; stats collected 2 minutes ago)
        table: nexus_service@nexus_service_pkey
        spans: FULL SCAN
(13 rows)


Time: 3ms total (execution 3ms / network 0ms)

```

and that's the plan I'd expect on a system with lots of external IPs allocated but not _that_ many Nexus instances.  Compare to:

```
root@127.0.0.1:60981/omicron> EXPLAIN SELECT * FROM external_ip WHERE id IN (SELECT external_ip_id from nexus_service);
                                         info
---------------------------------------------------------------------------------------
  distribution: local
  vectorized: true

  • lookup join
  │ estimated row count: 1
  │ table: external_ip@external_ip_pkey
  │ equality: (external_ip_id) = (id)
  │ equality cols are key
  │
  └── • distinct
      │ estimated row count: 1
      │ distinct on: external_ip_id
      │
      └── • scan
            estimated row count: 1 (100% of the table; stats collected 3 minutes ago)
            table: nexus_service@nexus_service_pkey
            spans: FULL SCAN

  index recommendations: 1
  1. type: index creation
     SQL command: CREATE INDEX ON nexus_service (external_ip_id);
(21 rows)


Time: 3ms total (execution 2ms / network 1ms)
```

This is almost the same plan except for the "distinct" node.  This made me realize that the INNER JOIN will show duplicates if there are two Nexus instances with the same IP, whereas the current query won't.  Today I think we expect this won't ever happen.  I wonder if that would be useful, though? Today, we can't express that external DNS and Nexus are both on 127.0.0.1 in development and I haven't figured out how we're going to resolve that.  Maybe one way is to allow two services to share an `external_ip`?  I dunno and that's probably beyond the scope of this PR.

Anyway, if we _do_ want this to be unique, we could add an index.  I did this wrong first:

```
root@127.0.0.1:60981/omicron> CREATE UNIQUE INDEX ON omicron.public.nexus_service (id, external_ip_id);
CREATE INDEX


Time: 115ms total (execution 115ms / network 0ms)

root@127.0.0.1:60981/omicron>
root@127.0.0.1:60981/omicron> EXPLAIN SELECT * FROM external_ip WHERE id IN (SELECT external_ip_id from nexus_service);
                                         info
---------------------------------------------------------------------------------------
  distribution: local
  vectorized: true

  • lookup join
  │ estimated row count: 1
  │ table: external_ip@external_ip_pkey
  │ equality: (external_ip_id) = (id)
  │ equality cols are key
  │
  └── • distinct
      │ estimated row count: 1
      │ distinct on: external_ip_id
      │
      └── • scan
            estimated row count: 1 (100% of the table; stats collected 7 minutes ago)
            table: nexus_service@nexus_service_pkey
            spans: FULL SCAN

  index recommendations: 1
  1. type: index creation
     SQL command: CREATE INDEX ON nexus_service (external_ip_id);
(21 rows)


Time: 3ms total (execution 2ms / network 0ms)

```

It would probably have to look like this:

```
root@127.0.0.1:60981/omicron> CREATE UNIQUE INDEX ON omicron.public.nexus_service (external_ip_id);
CREATE INDEX


Time: 101ms total (execution 101ms / network 0ms)

root@127.0.0.1:60981/omicron> EXPLAIN SELECT * FROM external_ip WHERE id IN (SELECT external_ip_id from nexus_service);
                                       info
-----------------------------------------------------------------------------------
  distribution: local
  vectorized: true

  • lookup join
  │ estimated row count: 1
  │ table: external_ip@external_ip_pkey
  │ equality: (external_ip_id) = (id)
  │ equality cols are key
  │
  └── • scan
        estimated row count: 1 (100% of the table; stats collected 2 minutes ago)
        table: nexus_service@nexus_service_pkey
        spans: FULL SCAN
(13 rows)


Time: 3ms total (execution 2ms / network 0ms)

root@127.0.0.1:60981/omicron>
```

and I think that _is_ the same plan as before.

---

I've never used the INNER JOIN syntax in production before, as I find the explicit use of conditions clearer for people unfamiliar with that syntax, but I see we're using it in plenty of places already so I'll take a look at doing that here.

Reply:
By explicit use of conditions do you mean without the `joinable!` macro? If so, that is what I was imagining with my suggestion 😅️ Up to you if you want to switch to that or leave as-is.